### ISSUE TITLE
`runc spec` failed on fedora 23 if `rootfs` is owned by a non-root user

### ISSUE THREAD

**Issue by hmeng-19** [opencontainers/runc#862]
I am running the master branch of `runc` on fedora 23. The `config.json` created by `runc spec` does not include `uid` and `gid` info (as shown below), therefore `runc start test` fails with an error message `permission denied`.

    "process": {
        "terminal": true,
        "user": {},
        "args": [
            "sh"
        ],

Proposed solution: add `uid` and `gid` info into the `config.json` file created by `runc spec` (as shown below). Here the uid:gid info is the uid:gid info of the rootfs specified in the `config.json`.

    "process": {
        "terminal": true,
        "user": {
            "uid": 1000,
            "gid": 1000,
            "additionalGids": null
        },
        "args": [
            "sh"
        ],

Notice, it seems that the `uid` and `gid` must match the uid:gid info of the rootfs. Otherwise, it still fails with `permission denied`. For example, if the rootfs is owned by a non-root user, and the uid:gid is set to be root (0:0) in `config.json`, `runc start test` would fail.

**Comment by cyphar** [comment]
How are you running `runc`? Since you're meant to run it as root (for now), there shouldn't be any permission denied issues. I believe the reason that we're not outputting the values for user is because the defaults we use are

```
"user": {
  "uid": 0,
  "gid": 0,
  "additionalGids": null
}
```

Which also happens to be the zero value for that structure -- so Go doesn't output the settings. But it is identical to specifying the `uid` and `gid` as root.

**Comment by hmeng-19** [comment]
My uid:gid is hmeng:hmeng (1000:1000), and I added this user info into the docker group on my machine.
 
I created the rootfs using the following command:
 
    hmeng@localhost ~/intern/runc/busybox$ docker export $(docker create busybox) | tar -C rootfs -xvf -
 
Here is the stat info of the rootfs:
 
    hmeng@localhost ~/intern/runc/busybox$ stat rootfs/
      File: ‘rootfs/’
      Size: 4096            Blocks: 8          IO Block: 4096   directory
    Device: fd02h/64770d    Inode: 535004      Links: 12
    Access: (0750/drwxr-x---)  Uid: ( 1000/   hmeng)   Gid: ( 1000/   hmeng)
    Context: unconfined_u:object_r:user_home_t:s0
    Access: 2016-05-31 22:21:55.900156998 -0400
    Modify: 2016-05-31 22:21:58.303143878 -0400
    Change: 2016-05-31 22:21:58.303143878 -0400
     Birth: -
 
Then I created the `config.json` by running the following command:
 
    hmeng@localhost ~/intern/runc/busybox$ /home/hmeng/go/src/github.com/hmeng-19/runc/runc spec
 
I ran `runc start test` and it failed due to the mismatch between the default uid:gid info used by runc and the ownership info of the rootfs:
 
    hmeng@localhost ~/intern/runc/busybox$ sudo /home/hmeng/go/src/github.com/hmeng-19/runc/runc start test
    permission denied
 
Then I changed the owner of the rootfs to root, and reran runc command, it worked:
 
    hmeng@localhost ~/intern/runc/busybox$ sudo chown -R root:root rootfs/
    hmeng@localhost ~/intern/runc/busybox$ sudo /home/hmeng/go/src/github.com/hmeng-19/runc/runc start test

**Comment by hmeng-19** [comment]
I think you are right about the default value of uid:gid info (0:0). 

The problem here seems to be that runc expects the rootfs to be owned by root. However, if a non-root user, `user1`, under the docker group, creates the rootfs using `docker export`, the owner of the rootfs becomes `user1`, which is not expected by runc.

BTW, do you think it is a good idea to check the ownership of the rootfs when executing `runc spec`, and put the owner info into the `config.json`.

**Comment by cyphar** [comment]
I would prefer not to add checks like that to `runc spec` (and not change `runc spec` at all if possible). I actually would want to fix the underlying issue with having the `rootfs` be owned by a different user even though we're root -- that looks to be the underlying bug. If you agree, can you change the title to reflect that?

**Comment by hmeng-19** [comment]
@cyphar , thanks for your help.
However, I still suffer the problem. 
If you can not reproduce the problem on opensuse, I would suspect this may be a problem special to fedora.
I would first try to reproduce this problem on another fedora machine.
Another problem with fedora is when you call `runc start test`, you must use the full path of runc like `sudo /usr/local/bin/runc start test` (http://echorand.me/runc-and-libcontainer-on-fedora-2324.html#.V07pREorKV4).

**Comment by runcom** [comment]
> Another problem with fedora is when you call runc start test, you must use the full path of runc like sudo /usr/local/bin/runc start test (http://echorand.me/runc-and-libcontainer-on-fedora-2324.html#.V07pREorKV4).

that's probably because you didn't run `sudo make install` and the runc binary is in dir of the project. Basically it's not in your `$PATH`

**Comment by cyphar** [comment]
@hmeng-19 Can you give us the output of (you might want to save it to a file):

```
% sudo strace -f runc start test
```

I'm also a bit annoyed that we don't output a more descriptive error message wherever we're failing (I thought we fixed this already).

**Comment by hmeng-19** [comment]
@runcom , It does not look like a problem with `$PATH`.
 
 If I run as a non-root user, `which runc` succeeds.

    hmeng@localhost ~/intern/runc/test$ which runc
    /usr/local/sbin/runc
 
However, `sudo which runc` failed.

    hmeng@localhost ~/intern/runc/test$ sudo which runc
    which: no runc in (/sbin:/bin:/usr/sbin:/usr/bin)
 
But, The `$PATH` env of the root user has `/usr/local/sbin`.
 
    hmeng@localhost ~/intern/runc/test$ sudo echo $PATH
    /usr/lib64/qt-3.3/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/usr/lib/golang/bin:/home/hmeng/go/bin
 
If I run as the root user, `which runc` has no problem.

    [root@localhost hmeng]# which runc
    /usr/local/sbin/runc

The kernel I am using is:

    [root@localhost hmeng]# uname -a
    Linux localhost.localdomain 4.4.9-300.fc23.x86_64 #1 SMP Wed May 4 23:56:27 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

**Comment by wking** [comment]
On Wed, Jun 01, 2016 at 08:06:55AM -0700, hmeng-19 wrote:
>     hmeng@localhost ~/intern/runc/test$ sudo echo $PATH
>     /usr/lib64/qt-3.3/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/usr/lib/golang/bin:/home/hmeng/go/bin

You non-root shell is expanding that PATH.  You want:

  $ sudo sh -c 'echo $PATH'

or some such.

**Comment by hmeng-19** [comment]
@wking , thanks for the explanation. I did not realize the command substitution and environment expansion happen before the root user gets the control.

**Comment by hmeng-19** [comment]
@runcom , @wking , I do not think this is a bug anymore. 

The summary is that, on Fedora 23, 
If `runc` is installed by `dnf install runc`, it is installed at `/usr/bin/runc`, and `/usr/bin` is included in the `$PATH` of root by default. Therefore, `sudo runc start test` works.

However, if `runc` is installed via building the source code, it is installed at `/usr/local/sbin/runc` by running `sudo make install`, and `/usr/local/sbin` is not in the `$PATH` of the root by default. Therefore, you must run `sudo /usr/local/sbin/runc start test`.

**Comment by hmeng-19** [comment]
@cyphar , I failed to reproduce the problem with the ownership of rootfs on other fedora 23 machines. 
Then I reinstalled fedora 23 on my own machine, and the problem disappeared. 

Unfortunately, I did not figure out why I suffered the ownership problem at the first place (that is why I reinstalled my system). My best guess is that my bash configurations may get messed up in the last two days. 

Thanks for your help. Sorry about the drama.
